File: sdk/typescript/src/tx.ts
/**
 * Unsigned transaction builder for Viper PQ Chain.
 *
 * SIGNING LIMITATION: ML-DSA and SLH-DSA (FIPS 204/205) have no mature
 * JavaScript/TypeScript implementation as of 2026. This builder produces
 * a structured unsigned transaction object. Signing must be performed
 * by the `pqcd sign-tx` CLI command or a Rust-based signing service.
 *
 * Workflow:
 *   1. Build unsigned tx with this module → get JSON payload.
 *   2. Pass JSON to `pqcd sign-tx --tx-json <file> --key-file <key>`.
 *   3. Submit the signed CBOR hex via ViperClient.submitTx().
 *
 * The "op_payload" field is a plain object; CBOR encoding is performed
 * by the pqcd CLI or server-side. This keeps the SDK dependency-free.
 */

import {
  AttestationCreateParams,
  TxOpType,
  ValidatorRegisterParams,
  VaultCreateParams,
  VaultTransferParams,
} from "./types.js";
import { assertValidAddress, hexToBytes } from "./utils.js";

export interface UnsignedTransaction {
  version: 1;
  op_type: TxOpType;
  sender: string;
  nonce: number;
  op_payload: Record<string, unknown>;
  /** Fee budget in venom as a decimal string. Must cover estimated fee. */
  fee_budget_venom: string;
  alg_id: string;
}

// ---------------------------------------------------------------------------
// Canonical token_transfer payload CBOR encoder
// ---------------------------------------------------------------------------
//
// The on-wire payload for a `token_transfer` tx is a CBOR map:
//   {1: recipient (bstr, 32 bytes), 2: amount}
//
// Amount (key 2) has two canonical encodings, matching the Rust encoder in
// `crates/pqcd/src/main.rs` (cmd_wallet_send) and the decoder's `expect_u128`
// in `crates/pqc-state/src/apply/transfer.rs`:
//   - CBOR unsigned integer (major type 0)  when amount <= u64::MAX
//   - CBOR byte string (major type 2, 16B)  when amount  > u64::MAX
//
// The bstr branch mirrors the u128 convention used for balances in
// `pqc_types::multisig::MultisigAccountState::to_cbor_bytes` — 16 bytes,
// big-endian. Amounts are `bigint` throughout; `number` only carries 53
// integer bits, nowhere near the 128-bit range.

const U64_MAX = (1n << 64n) - 1n;
const U128_MAX = (1n << 128n) - 1n;

/**
 * Encode an unsigned integer head (major type + argument) per RFC 8949 §3.
 * Emits the shortest form for the given value.
 */
function encodeUnsignedHead(majorType: number, value: bigint): Uint8Array {
  if (value < 0n) {
    throw new Error(`CBOR head argument must be non-negative, got ${value}`);
  }
  const mt = (majorType & 0x07) << 5;
  if (value < 24n) {
    return new Uint8Array([mt | Number(value)]);
  }
  if (value <= 0xffn) {
    return new Uint8Array([mt | 24, Number(value)]);
  }
  if (value <= 0xffffn) {
    const v = Number(value);
    return new Uint8Array([mt | 25, (v >>> 8) & 0xff, v & 0xff]);
  }
  if (value <= 0xffffffffn) {
    const v = Number(value);
    return new Uint8Array([
      mt | 26,
      (v >>> 24) & 0xff,
      (v >>> 16) & 0xff,
      (v >>> 8) & 0xff,
      v & 0xff,
    ]);
  }
  if (value <= U64_MAX) {
    const buf = new Uint8Array(9);
    buf[0] = mt | 27;
    let v = value;
    for (let i = 8; i >= 1; i--) {
      buf[i] = Number(v & 0xffn);
      v >>= 8n;
    }
    return buf;
  }
  throw new Error(`CBOR head argument exceeds u64: ${value}`);
}

/** Encode a CBOR byte string (major type 2). */
function encodeBytes(data: Uint8Array): Uint8Array {
  const head = encodeUnsignedHead(2, BigInt(data.length));
  const out = new Uint8Array(head.length + data.length);
  out.set(head, 0);
  out.set(data, head.length);
  return out;
}

/** Encode a CBOR unsigned integer (major type 0). */
function encodeUint(value: bigint): Uint8Array {
  if (value < 0n) {
    throw new Error(`CBOR uint must be non-negative, got ${value}`);
  }
  if (value <= U64_MAX) {
    return encodeUnsignedHead(0, value);
  }
  throw new Error(`CBOR uint exceeds u64 (${U64_MAX}): ${value}`);
}

/** Concatenate Uint8Array chunks. */
function concatBytes(chunks: Uint8Array[]): Uint8Array {
  let total = 0;
  for (const c of chunks) total += c.length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) {
    out.set(c, off);
    off += c.length;
  }
  return out;
}

/** Convert a u128 bigint to a 16-byte big-endian Uint8Array. */
function u128ToBeBytes(value: bigint): Uint8Array {
  if (value < 0n || value > U128_MAX) {
    throw new Error(`value out of u128 range: ${value}`);
  }
  const out = new Uint8Array(16);
  let v = value;
  for (let i = 15; i >= 0; i--) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

/**
 * Encode a `token_transfer` payload as CBOR bytes.
 *
 * The output matches the Rust encoder in `cmd_wallet_send`
 * (`crates/pqcd/src/main.rs`) and is round-trip compatible with the
 * decoder in `crates/pqc-state/src/apply/transfer.rs`.
 *
 * @param recipient - 32-byte recipient address, as a `Uint8Array` or a
 *                    64-char hex string.
 * @param amount    - Amount in venom as a `bigint`. Must fit in u128
 *                    (0..=2**128 - 1). Amounts up to `u64::MAX` are encoded
 *                    as a CBOR unsigned integer; larger amounts are encoded
 *                    as a 16-byte big-endian bytestring.
 */
export function encodeTokenTransferPayload(
  recipient: Uint8Array | string,
  amount: bigint
): Uint8Array {
  const recipientBytes =
    typeof recipient === "string" ? hexToBytes(recipient) : recipient;
  if (recipientBytes.length !== 32) {
    throw new Error(
      `recipient must be exactly 32 bytes, got ${recipientBytes.length}`
    );
  }
  if (typeof amount !== "bigint") {
    throw new TypeError(
      `amount must be a bigint; received ${typeof amount}. ` +
        `Use a BigInt literal (e.g. 1000n) to avoid JS Number precision loss.`
    );
  }
  if (amount < 0n) {
    throw new Error(`amount must be non-negative, got ${amount}`);
  }
  if (amount > U128_MAX) {
    throw new Error(`amount exceeds u128 range: ${amount}`);
  }

  // Map header: 2 entries → 0xA2.
  const mapHeader = new Uint8Array([0xa2]);

  // key 1 → recipient bytes
  const key1 = encodeUint(1n);
  const recipientCbor = encodeBytes(recipientBytes);

  // key 2 → amount (integer when ≤ u64::MAX, else 16-byte bstr)
  const key2 = encodeUint(2n);
  const amountCbor =
    amount <= U64_MAX ? encodeUint(amount) : encodeBytes(u128ToBeBytes(amount));

  return concatBytes([mapHeader, key1, recipientCbor, key2, amountCbor]);
}

/** Validate nonce is a non-negative integer. */
function assertNonce(nonce: number): void {
  if (!Number.isInteger(nonce) || nonce < 0) {
    throw new Error(`nonce must be a non-negative integer, got ${nonce}`);
  }
}

/** Build an unsigned vault_create transaction. */
export function buildVaultCreate(
  params: VaultCreateParams,
  feeBudgetVenom: bigint
): UnsignedTransaction {
  assertValidAddress(params.sender, "sender");
  assertNonce(params.nonce);
  if (!params.public_key) throw new Error("public_key is required");

  return {
    version: 1,
    op_type: "vault_create",
    sender: params.sender,
    nonce: params.nonce,
    op_payload: {
      alg_id: params.alg_id,
      public_key: params.public_key,
    },
    fee_budget_venom: feeBudgetVenom.toString(),
    alg_id: params.alg_id,
  };
}

/** Build an unsigned vault_transfer transaction. */
export function buildVaultTransfer(
  params: VaultTransferParams,
  feeBudgetVenom: bigint
): UnsignedTransaction {
  assertValidAddress(params.sender, "sender");
  assertValidAddress(params.recipient, "recipient");
  assertNonce(params.nonce);
  if (params.amount_venom <= 0n) {
    throw new Error("amount_venom must be positive");
  }

  return {
    version: 1,
    op_type: "vault_transfer",
    sender: params.sender,
    nonce: params.nonce,
    op_payload: {
      recipient: params.recipient,
      amount_venom: params.amount_venom.toString(),
    },
    fee_budget_venom: feeBudgetVenom.toString(),
    alg_id: "ml-dsa-65",
  };
}

/** Build an unsigned attestation_create transaction. */
export function buildAttestationCreate(
  params: AttestationCreateParams,
  feeBudgetVenom: bigint
): UnsignedTransaction {
  assertValidAddress(params.sender, "sender");
  assertValidAddress(params.subject, "subject");
  assertNonce(params.nonce);

  return {
    version: 1,
    op_type: "attestation_create",
    sender: params.sender,
    nonce: params.nonce,
    op_payload: {
      subject: params.subject,
      schema_id: params.schema_id,
      payload_hex: params.payload_hex,
    },
    fee_budget_venom: feeBudgetVenom.toString(),
    alg_id: "ml-dsa-65",
  };
}

/** Build an unsigned validator_register transaction. */
export function buildValidatorRegister(
  params: ValidatorRegisterParams,
  feeBudgetVenom: bigint
): UnsignedTransaction {
  assertValidAddress(params.sender, "sender");
  assertNonce(params.nonce);
  if (params.self_bond_venom <= 0n) {
    throw new Error("self_bond_venom must be positive");
  }

  return {
    version: 1,
    op_type: "validator_register",
    sender: params.sender,
    nonce: params.nonce,
    op_payload: {
      consensus_pk: params.consensus_pk,
      consensus_alg: params.consensus_alg,
      self_bond_venom: params.self_bond_venom.toString(),
    },
    fee_budget_venom: feeBudgetVenom.toString(),
    alg_id: "ml-dsa-65",
  };
}

/** Build an unsigned validator_exit transaction. */
export function buildValidatorExit(
  sender: string,
  nonce: number,
  feeBudgetVenom: bigint
): UnsignedTransaction {
  assertValidAddress(sender, "sender");
  assertNonce(nonce);

  return {
    version: 1,
    op_type: "validator_exit",
    sender,
    nonce,
    op_payload: {},
    fee_budget_venom: feeBudgetVenom.toString(),
    alg_id: "ml-dsa-65",
  };
}

File: sdk/python/viper_pqchain/tx.py
"""
Unsigned transaction builder for Viper PQ Chain.

SIGNING LIMITATION: ML-DSA and SLH-DSA (FIPS 204/205) have no mature Python
implementation as of 2026. This builder produces a structured unsigned transaction
dictionary. Signing must be performed by the ``pqcd sign-tx`` CLI command or a
Rust-based signing service.

Workflow:
  1. Build unsigned tx with this module → get a dict payload.
  2. Serialise to JSON and pass to ``pqcd sign-tx --tx-json <file> --key-file <key>``.
  3. Submit the signed CBOR hex via ``ViperClient.submit_tx()``.

The ``op_payload`` field is a plain dict; CBOR encoding is performed by the
pqcd CLI or server-side. This keeps the SDK dependency-free.
"""

from __future__ import annotations

from typing import Any, Union

from .types import (
    AttestationCreateParams,
    ValidatorRegisterParams,
    VaultCreateParams,
    VaultTransferParams,
)
from .utils import assert_nonce, assert_valid_address

UnsignedTransaction = dict[str, Any]


# ---------------------------------------------------------------------------
# Canonical token_transfer payload CBOR encoder
# ---------------------------------------------------------------------------
#
# The on-wire payload for a ``token_transfer`` tx is a CBOR map:
#   {1: recipient (bstr, 32 bytes), 2: amount}
#
# Amount (key 2) has two canonical encodings, matching the Rust encoder in
# ``crates/pqcd/src/main.rs`` (``cmd_wallet_send``) and the decoder's
# ``expect_u128`` in ``crates/pqc-state/src/apply/transfer.rs``:
#   - CBOR unsigned integer (major type 0)  when amount <= u64::MAX
#   - CBOR byte string (major type 2, 16B)  when amount  > u64::MAX
#
# The bstr branch mirrors the u128 convention used for balances in
# ``pqc_types::multisig::MultisigAccountState::to_cbor_bytes`` — 16 bytes,
# big-endian.

_U64_MAX = (1 << 64) - 1
_U128_MAX = (1 << 128) - 1


def _encode_unsigned_head(major_type: int, value: int) -> bytes:
    """Emit a CBOR (major_type, argument) head per RFC 8949 §3, shortest form."""
    if value < 0:
        raise ValueError(f"CBOR head argument must be non-negative, got {value}")
    mt = (major_type & 0x07) << 5
    if value < 24:
        return bytes([mt | value])
    if value <= 0xFF:
        return bytes([mt | 24, value])
    if value <= 0xFFFF:
        return bytes([mt | 25]) + value.to_bytes(2, "big")
    if value <= 0xFFFFFFFF:
        return bytes([mt | 26]) + value.to_bytes(4, "big")
    if value <= _U64_MAX:
        return bytes([mt | 27]) + value.to_bytes(8, "big")
    raise ValueError(f"CBOR head argument exceeds u64: {value}")


def _encode_bytes(data: bytes) -> bytes:
    """Encode a CBOR byte string (major type 2)."""
    return _encode_unsigned_head(2, len(data)) + data


def _encode_uint(value: int) -> bytes:
    """Encode a CBOR unsigned integer (major type 0)."""
    if value < 0:
        raise ValueError(f"CBOR uint must be non-negative, got {value}")
    if value > _U64_MAX:
        raise ValueError(f"CBOR uint exceeds u64 ({_U64_MAX}): {value}")
    return _encode_unsigned_head(0, value)


def encode_token_transfer_payload(
    recipient: Union[bytes, bytearray, memoryview, str],
    amount: int,
) -> bytes:
    """Encode a ``token_transfer`` payload as CBOR bytes.

    The output matches the Rust encoder in ``cmd_wallet_send``
    (``crates/pqcd/src/main.rs``) and is round-trip compatible with the
    decoder in ``crates/pqc-state/src/apply/transfer.rs``.

    :param recipient: 32-byte recipient address. Accepted as ``bytes``,
        ``bytearray``, ``memoryview``, or a 64-char hex string.
    :param amount: Amount in venom as a Python ``int``. Must fit in u128
        (``0..=2**128 - 1``). Amounts up to ``u64::MAX`` are encoded as a
        CBOR unsigned integer; larger amounts are encoded as a 16-byte
        big-endian bytestring.
    :raises TypeError: if *amount* is not an ``int`` (e.g. a ``bool`` or
        ``float``). ``bool`` is rejected explicitly because it is an ``int``
        subclass but semantically unsafe here.
    :raises ValueError: if *recipient* is not exactly 32 bytes or *amount*
        is outside ``[0, 2**128 - 1]``.
    """
    if isinstance(recipient, str):
        clean = recipient.removeprefix("0x").lower()
        recipient_bytes = bytes.fromhex(clean)
    else:
        recipient_bytes = bytes(recipient)

    if len(recipient_bytes) != 32:
        raise ValueError(
            f"recipient must be exactly 32 bytes, got {len(recipient_bytes)}"
        )

    # bool is an int subclass in Python; reject it to catch accidental True/False.
    if isinstance(amount, bool) or not isinstance(amount, int):
        raise TypeError(
            f"amount must be an int; received {type(amount).__name__}"
        )
    if amount < 0:
        raise ValueError(f"amount must be non-negative, got {amount}")
    if amount > _U128_MAX:
        raise ValueError(f"amount exceeds u128 range: {amount}")

    # Map header: 2 entries → 0xA2.
    map_header = bytes([0xA2])

    # key 1 → recipient bstr(32)
    key1 = _encode_uint(1)
    recipient_cbor = _encode_bytes(recipient_bytes)

    # key 2 → amount (integer when ≤ u64::MAX, else 16-byte bstr)
    key2 = _encode_uint(2)
    if amount <= _U64_MAX:
        amount_cbor = _encode_uint(amount)
    else:
        amount_cbor = _encode_bytes(amount.to_bytes(16, "big"))

    return map_header + key1 + recipient_cbor + key2 + amount_cbor


def build_vault_create(
    params: VaultCreateParams,
    fee_budget_venom: int,
) -> UnsignedTransaction:
    """Build an unsigned vault_create transaction."""
    assert_valid_address(params.sender, "sender")
    assert_nonce(params.nonce)
    if not params.public_key:
        raise ValueError("public_key is required")

    return {
        "version": 1,
        "op_type": "vault_create",
        "sender": params.sender,
        "nonce": params.nonce,
        "op_payload": {
            "alg_id": params.alg_id,
            "public_key": params.public_key,
        },
        "fee_budget_venom": str(fee_budget_venom),
        "alg_id": params.alg_id,
    }


def build_vault_transfer(
    params: VaultTransferParams,
    fee_budget_venom: int,
) -> UnsignedTransaction:
    """Build an unsigned vault_transfer transaction."""
    assert_valid_address(params.sender, "sender")
    assert_valid_address(params.recipient, "recipient")
    assert_nonce(params.nonce)
    if params.amount_venom <= 0:
        raise ValueError("amount_venom must be positive")

    return {
        "version": 1,
        "op_type": "vault_transfer",
        "sender": params.sender,
        "nonce": params.nonce,
        "op_payload": {
            "recipient": params.recipient,
            "amount_venom": str(params.amount_venom),
        },
        "fee_budget_venom": str(fee_budget_venom),
        "alg_id": "ml-dsa-65",
    }


def build_attestation_create(
    params: AttestationCreateParams,
    fee_budget_venom: int,
) -> UnsignedTransaction:
    """Build an unsigned attestation_create transaction."""
    assert_valid_address(params.sender, "sender")
    assert_valid_address(params.subject, "subject")
    assert_nonce(params.nonce)

    return {
        "version": 1,
        "op_type": "attestation_create",
        "sender": params.sender,
        "nonce": params.nonce,
        "op_payload": {
            "subject": params.subject,
            "schema_id": params.schema_id,
            "payload_hex": params.payload_hex,
        },
        "fee_budget_venom": str(fee_budget_venom),
        "alg_id": "ml-dsa-65",
    }


def build_validator_register(
    params: ValidatorRegisterParams,
    fee_budget_venom: int,
) -> UnsignedTransaction:
    """Build an unsigned validator_register transaction."""
    assert_valid_address(params.sender, "sender")
    assert_nonce(params.nonce)
    if params.self_bond_venom <= 0:
        raise ValueError("self_bond_venom must be positive")

    return {
        "version": 1,
        "op_type": "validator_register",
        "sender": params.sender,
        "nonce": params.nonce,
        "op_payload": {
            "consensus_pk": params.consensus_pk,
            "consensus_alg": params.consensus_alg,
            "self_bond_venom": str(params.self_bond_venom),
        },
        "fee_budget_venom": str(fee_budget_venom),
        "alg_id": "ml-dsa-65",
    }


def build_validator_exit(
    sender: str,
    nonce: int,
    fee_budget_venom: int,
) -> UnsignedTransaction:
    """Build an unsigned validator_exit transaction."""
    assert_valid_address(sender, "sender")
    assert_nonce(nonce)

    return {
        "version": 1,
        "op_type": "validator_exit",
        "sender": sender,
        "nonce": nonce,
        "op_payload": {},
        "fee_budget_venom": str(fee_budget_venom),
        "alg_id": "ml-dsa-65",
    }
